### ISSUE TITLE
Kubelet cannot start static pod when move in manifest file after removing manifest file for a while

### ISSUE THREAD

**Issue by liushi001** [kubernetes/kubernetes#107483]
### What happened?

An error occurs in `managePodLoop` after the static pod has being deleted, then `syncCh` is triggerred and error event will be dequeued from work queue but pod manage has no such pod because `HandlePodRemoves` has been called.

`No such container error` is a common error type from `p.podCache.GetNewerThan` or `syncTerminatingPodFn` method in `managePodLoop` .

So this event can not trigger `HandlePodSyncs` to finish pod terminating.

Because of terminating state of pod, `SyncLoop ADD` will mark restartable flag and return.

So the housekeeper becomes the only way to start the pod. But also due to terminating state(not finished), the pod cannot be added to `restartablePods`, so it cannot be retarted again.

Eventually, the pod will never be restarted again.

### What did you expect to happen?

Static pod can be started or restarted again.

### How can we reproduce it (as minimally and precisely as possible)?

We found that the pod would fail to start after the events occurred in the following order:
1. HandlePodRemoves -> podManager.DeletePod -> managePodLoop ->  err occurred, like "No such container" -> completeWork -> workQueue.Enqueue

2. syncCh -> getPodsToSync -> workQueue.GetWork() -> podManager.GetPods() -> pod has been deleted, and HandlePodSyncs cannot be entered 

3. HandlePodAdditions -> podManager.AddPod -> status.IsTerminationRequested -> status.restartRequested=true -> return

4. HandlePodCleanups -> podWorkers.SyncKnownPods -> removeTerminatedWorker -> !status.finished -> do not delete podSyncStatuses -> status.terminatedAt.IsZero() && !status.terminatingAt.IsZero() -> still in TerminatingPod state

Then the pod will never be started due to lack of trigger of pod worker, unless the kubelet is restarted.

Stub method for error in managePodLoop:
1. Add a delay between `podSandboxIDs, err := m.getSandboxIDByPodUID(uid, nil)` and `podSandboxStatus, err := m.runtimeService.PodSandboxStatus(podSandboxID)` in for range in `GetPodStatus` in `syncTerminatingPod` .
2. Once `getSandboxIDByPodUID` has been called, kill and remove the pause container of the pod manually.
3. An `No such container` error occurs.

Kubelet Log:
<details>

```log
I1213 17:26:45.869540   19130 kubelet.go:2249] "SyncLoop REMOVE" source="file" pods=[fst-manage/webhook-paas-192-168-2-2]
I1213 17:26:45.869806   19130 kuberuntime_container.go:721] "Killing container with a grace period" pod="fst-manage/webhook-paas-192-168-2-2" podUID=3a91011824b5370050c4bb2c2c0705ad containerName="webhook" containerID="docker://9ff7d0fbdb1f96db833ae273cee55cec2ae73cdd201088d9b6af515c2879d350" gracePeriod=5
I1213 17:26:46.211320   19130 reconciler.go:197] "operationExecutor.UnmountVolume started for volume \"localtime\" (UniqueName: \"kubernetes.io/host-path/3a91011824b5370050c4bb2c2c0705ad-localtime\") pod \"3a91011824b5370050c4bb2c2c0705ad\" (UID: \"3a91011824b5370050c4bb2c2c0705ad\") "
...
I1213 17:26:46.761943   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="9ff7d0fbdb1f96db833ae273cee55cec2ae73cdd201088d9b6af515c2879d350" oldState=running newState=exited
I1213 17:26:46.761949   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="5b0e3b74c2a73880c3e984fac3c40c8ef0648c92b3df6c0cbf178d4541af5e2d" oldState=running newState=exited
E1213 17:26:47.292895   19130 kuberuntime_manager.go:1091] "getPodContainerStatuses for pod failed" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2"
E1213 17:26:47.292948   19130 pod_workers.go:894] "Error syncing pod, skipping" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2" podUID=3a91011824b5370050c4bb2c2c0705ad
I1213 17:26:48.312614   19130 kubelet_volumes.go:161] "Cleaned up orphaned pod volumes dir" podUID=3a91011824b5370050c4bb2c2c0705ad path="/var/lib/kubelet/pods/3a91011824b5370050c4bb2c2c0705ad/volumes"
I1213 17:26:48.712057   19130 kubelet.go:2281] "SyncLoop (PLEG): pod does not exist, ignore irrelevant event" event=&{ID:3a91011824b5370050c4bb2c2c0705ad Type:ContainerDied Data:5b0e3b74c2a73880c3e984fac3c40c8ef0648c92b3df6c0cbf178d4541af5e2d}
I1213 17:27:00.876529   19130 kubelet.go:2239] "SyncLoop ADD" source="file" pods=[fst-manage/webhook-paas-192-168-2-2]
```

Another reoccurrence:
```log
I0110 11:56:03.480296   25160 kubelet.go:2249] "SyncLoop REMOVE" source="file" pods=[fst-manage/kube-apiserver-paas-192-168-2-2]
I0110 11:56:03.480433   25160 pod_workers.go:840] "Got non-process pod, start sync pod" pod="fst-manage/kube-apiserver-paas-192-168-2-2"
I0110 11:56:03.480768   25160 kuberuntime_container.go:721] "Killing container with a grace period" pod="fst-manage/kube-apiserver-paas-192-168-2-2" podUID=c6dc565f27a76405ed35d58bc441a407 containerName="kube-apiserver" containerID="docker://524779760c0050c8391419ea54295020f42460802cea89a2f11df8d08cb87f15" gracePeriod=1
I0110 11:56:05.297011   25160 generic.go:159] "GenericPLEG" podUID=c6dc565f27a76405ed35d58bc441a407 containerID="524779760c0050c8391419ea54295020f42460802cea89a2f11df8d08cb87f15" oldState=running newState=exited
I0110 11:56:06.177361   25160 generic.go:338] "Generic (PLEG): container finished" podID=c6dc565f27a76405ed35d58bc441a407 containerID="524779760c0050c8391419ea54295020f42460802cea89a2f11df8d08cb87f15" exitCode=137
I0110 11:56:06.177414   25160 kubelet.go:2281] "SyncLoop (PLEG): pod does not exist, ignore irrelevant event" event=&{ID:c6dc565f27a76405ed35d58bc441a407 Type:ContainerDied Data:524779760c0050c8391419ea54295020f42460802cea89a2f11df8d08cb87f15}
E0110 11:56:07.147581   25160 kuberuntime_manager.go:1091] "getPodContainerStatuses for pod failed" err="Error: No such container: fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" pod="fst-manage/kube-apiserver-paas-192-168-2-2"
E0110 11:56:07.147608   25160 kubelet.go:1983] "Unable to read pod status prior to final pod termination" err="Error: No such container: fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" pod="fst-manage/kube-apiserver-paas-192-168-2-2" podUID=c6dc565f27a76405ed35d58bc441a407
E0110 11:56:07.147629   25160 pod_workers.go:894] "Error syncing pod, skipping" err="Error: No such container: fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" pod="fst-manage/kube-apiserver-paas-192-168-2-2" podUID=c6dc565f27a76405ed35d58bc441a407
I0110 11:56:07.245849   25160 generic.go:159] "GenericPLEG" podUID=c6dc565f27a76405ed35d58bc441a407 containerID="fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" oldState=exited newState=non-existent
I0110 11:56:07.245860   25160 generic.go:159] "GenericPLEG" podUID=c6dc565f27a76405ed35d58bc441a407 containerID="349f55d33dc68eddebc3383a251e8048ceaa83cf0b006843b99e168736e61b82" oldState=running newState=exited
I0110 11:56:07.257231   25160 kubelet.go:2281] "SyncLoop (PLEG): pod does not exist, ignore irrelevant event" event=&{ID:c6dc565f27a76405ed35d58bc441a407 Type:ContainerDied Data:349f55d33dc68eddebc3383a251e8048ceaa83cf0b006843b99e168736e61b82}
I0110 11:57:20.569359   25160 kubelet.go:2239] "SyncLoop ADD" source="file" pods=[fst-manage/kube-apiserver-paas-192-168-2-2]
```
</details>

### Kubernetes version

<details>

```console
$ kubectl version
Kubernetes version: v1.22.1-c5fe23a7fbdd1d996731d6c996ba8aed3804d471
and already backportted  #105075
```
</details>

### Cloud provider

<details>

Local machine
</details>

### OS version

<details>

```console  
$ cat /etc/os-release
NAME="EulerOS"
VERSION="2.0 (SP10x86_64)"
ID="euleros"
VERSION_ID="2.0"
PRETTY_NAME="EulerOS 2.0 (SP10x86_64)"
ANSI_COLOR="0;31"
$ uname -a
Linux host-192-168-2-2 4.18.0-147.5.2.5.h732.eulerosv2r10.x86_64 #1 SMP Sat Nov 27 16:33:23 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux  
```
</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by 249043822** [comment]
this pr have fixed the problem, https://github.com/kubernetes/kubernetes/pull/105075
please upgrade to 1.22.3

**Comment by yangjunmyfm192085** [comment]
we will close this issue, thanks for your feedback.
/close

**Comment by liushi001** [comment]
sorry, I lost some infomation, I'v already backportted this pr #105075 just to resolved moving out and in static pod manifest file to restart pod in a short time.
But there happend another problem in the scenario than there is a long period of time before files are moved in and removed.
I think my problem just was introduced by this pr #105075.
/reopen

**Comment by smarterclayton** [comment]
How long was the static pod definition removed for?

**Comment by smarterclayton** [comment]
Definitely syncTerminatingPod and syncTerminatedPod should not be failing due to "no such container" - if a container can't be found, by definition nothing needs to be done.  Looking.

**Comment by smarterclayton** [comment]
If this is triggered by

```
	// Guard against consistency issues in KillPod implementations by checking that there are no
	// running containers. This method is invoked infrequently so this is effectively free and can
	// catch race conditions introduced by callers updating pod status out of order.
	// TODO: have KillPod return the terminal status of stopped containers and write that into the
	//  cache immediately
	podStatus, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
```

the bug is that GetPodStatus must return an unambiguous error to this code "the pod has no running containers" that the method can use to determine "we're definitely stopped".  We can't exit syncTerminatingPod on an error - we have to keep retrying.  So if you can show me the log line that shows the "no such error" and we can confirm it's coming from GetPodStatus, that should be enough to run it down.  But that is ONLY if you are certain that the pod worker is still trying to run syncTerminatingPod and that's why the new pod isn't starting.

**Comment by liushi001** [comment]
> How long was the static pod definition removed for?

At least 15 seconds or more, as following log shows(1 minutes in another recurrence):

I1213 17:26:45.869540   19130 kubelet.go:2249] "SyncLoop REMOVE" source="file" pods=[fst-manage/webhook-paas-192-168-2-2]
...
I1213 17:27:00.876529   19130 kubelet.go:2239] "SyncLoop ADD" source="file" pods=[fst-manage/webhook-paas-192-168-2-2]

**Comment by liushi001** [comment]
> Definitely syncTerminatingPod and syncTerminatedPod should not be failing due to "no such container" - if a container can't be found, by definition nothing needs to be done. Looking.

I think "no such container" maybe generated from previous residue container, whose id is different from ones of life cycyle changed shown by PLEG log:

After killing container, logs print there are two conainter exited which are 9ff7d0f and 5b0e3b7. But in "no such container" error, the container ID is 740ab75. I think it maybe removed by GC just at the time between ListContainer and InspectConainer CRI call.

```log 
I1213 17:26:46.761943   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="9ff7d0fbdb1f96db833ae273cee55cec2ae73cdd201088d9b6af515c2879d350" oldState=running newState=exited
I1213 17:26:46.761949   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="5b0e3b74c2a73880c3e984fac3c40c8ef0648c92b3df6c0cbf178d4541af5e2d" oldState=running newState=exited
E1213 17:26:47.292895   19130 kuberuntime_manager.go:1091] "getPodContainerStatuses for pod failed" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2"
E1213 17:26:47.292948   19130 pod_workers.go:894] "Error syncing pod, skipping" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2" podUID=3a91011824b5370050c4bb2c2c0705ad
```

**Comment by liushi001** [comment]
> > Definitely syncTerminatingPod and syncTerminatedPod should not be failing due to "no such container" - if a container can't be found, by definition nothing needs to be done. Looking.
> 
> I think "no such container" maybe generated from previous residue container, whose id is different from ones of life cycyle changed shown by PLEG log:
> 
> After killing container, logs print there are two conainter exited which are 9ff7d0f and [5b0e3b7](https://github.com/kubernetes/kubernetes/commit/5b0e3b75a43313ce8142bd4613bf7a0823d30ad2). But in "no such container" error, the container ID is 740ab75. I think it maybe removed by GC just at the time between ListContainer and InspectConainer CRI call.
> 
> ```
> I1213 17:26:46.761943   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="9ff7d0fbdb1f96db833ae273cee55cec2ae73cdd201088d9b6af515c2879d350" oldState=running newState=exited
> I1213 17:26:46.761949   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="5b0e3b74c2a73880c3e984fac3c40c8ef0648c92b3df6c0cbf178d4541af5e2d" oldState=running newState=exited
> E1213 17:26:47.292895   19130 kuberuntime_manager.go:1091] "getPodContainerStatuses for pod failed" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2"
> E1213 17:26:47.292948   19130 pod_workers.go:894] "Error syncing pod, skipping" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2" podUID=3a91011824b5370050c4bb2c2c0705ad
> ```



> If this is triggered by this method in `syncTerminatingPod` returning a `no such container` error:
> 
> ```
> 	// Guard against consistency issues in KillPod implementations by checking that there are no
> 	// running containers. This method is invoked infrequently so this is effectively free and can
> 	// catch race conditions introduced by callers updating pod status out of order.
> 	// TODO: have KillPod return the terminal status of stopped containers and write that into the
> 	//  cache immediately
> 	podStatus, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
> ```
> 
> the bug is that GetPodStatus must return an unambiguous error when "the pod has no running containers" that `syncTerminatingPod` can use to determine "we're definitely stopped and can ignore this error". We can't exit syncTerminatingPod on an error - we have to keep retrying. So if you can show me the log line that shows the "no such error" and we can confirm it's coming from GetPodStatus, that should be enough to identify where the error is getting wrapped. But that is ONLY if you are certain that the pod worker is still trying to run syncTerminatingPod and that's why the new pod isn't starting.
> 
> So log lines showing the error, and confirming that you are seeing the pod worker stuck in syncTerminatingPod (again, log snippets would be helpful) and we can run it down more.

For details about logs, please see the collapsed `Kubelet Log` part above. I'll pick key lines and explain where is comes from:

I confirmed that the first time the problem occurs, "no such container error" is from this method in `managePodLoop`:
```go
 // wait until we see the next refresh from the PLEG via the cache (max 2s)
 // TODO: this adds ~1s of latency on all transitions from sync to terminating
 //  to terminated, and on all termination retries (including evictions). We should
 //  improve latency by making the the pleg continuous and by allowing pod status
 //  changes to be refreshed when key events happen (killPod, sync->terminating).
 //  Improving this latency also reduces the possibility that a terminated
 //  container's status is garbage collected before we have a chance to update the
 //  API server (thus losing the exit code).
 status, err = p.podCache.GetNewerThan(pod.UID, lastSyncTime)
```

In later problem recurs for the second time, it is from `GetPodStatus` in `syncTerminatingPod`:
```
// Guard against consistency issues in KillPod implementations by checking that there are no
// running containers. This method is invoked infrequently so this is effectively free and can
// catch race conditions introduced by callers updating pod status out of order.
// TODO: have KillPod return the terminal status of stopped containers and write that into the
//  cache immediately
 podStatus, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
```
The following logs shows the message `Unable to read pod status prior to final pod termination`  which was printed after an error is returned by method `GetPodStatus` .

```log
I0110 11:56:03.480296   25160 kubelet.go:2249] "SyncLoop REMOVE" source="file" pods=[fst-manage/kube-apiserver-paas-192-168-2-2]
...
E0110 11:56:07.147581   25160 kuberuntime_manager.go:1091] "getPodContainerStatuses for pod failed" err="Error: No such container: fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" pod="fst-manage/kube-apiserver-paas-192-168-2-2"
E0110 11:56:07.147608   25160 kubelet.go:1983] "Unable to read pod status prior to final pod termination" err="Error: No such container: fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" pod="fst-manage/kube-apiserver-paas-192-168-2-2" podUID=c6dc565f27a76405ed35d58bc441a407
E0110 11:56:07.147629   25160 pod_workers.go:894] "Error syncing pod, skipping" err="Error: No such container: fa6daeb414be6249f3f8002c7c6098bcf888debeebcef108d7c6b2fb5cfea447" pod="fst-manage/kube-apiserver-paas-192-168-2-2" podUID=c6dc565f27a76405ed35d58bc441a407
...
I0110 11:57:20.569359   25160 kubelet.go:2239] "SyncLoop ADD" source="file" pods=[fst-manage/kube-apiserver-paas-192-168-2-2]
```

**Comment by 249043822** [comment]
> > Definitely syncTerminatingPod and syncTerminatedPod should not be failing due to "no such container" - if a container can't be found, by definition nothing needs to be done. Looking.
> 
> After killing container, logs print there are two conainter exited which are `9ff7d0f` and `5b0e3b7`. But in "no such container" error, the container ID is `740ab75`. I think it maybe removed by GC just at the time between ListContainer and InspectConainer CRI call.
> 
> I think "no such container" maybe generated from previous residue container already exited, due to its ID is different from exited event life cycyle log shown above .
> 
> ```
> I1213 17:26:46.761943   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="9ff7d0fbdb1f96db833ae273cee55cec2ae73cdd201088d9b6af515c2879d350" oldState=running newState=exited
> I1213 17:26:46.761949   19130 generic.go:159] "GenericPLEG" podUID=3a91011824b5370050c4bb2c2c0705ad containerID="5b0e3b74c2a73880c3e984fac3c40c8ef0648c92b3df6c0cbf178d4541af5e2d" oldState=running newState=exited
> E1213 17:26:47.292895   19130 kuberuntime_manager.go:1091] "getPodContainerStatuses for pod failed" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2"
> E1213 17:26:47.292948   19130 pod_workers.go:894] "Error syncing pod, skipping" err="Error: No such container: 740ab753973f7ad32c5ef2c9778fb061469fc1a664710988569a46deb7eea66e" pod="fst-manage/webhook-paas-192-168-2-2" podUID=3a91011824b5370050c4bb2c2c0705ad
> ```

This analysis seems reasonable, actually in the gap for listcontainers and inspectcontainer in `getPodContainerStatuses  `, containergc can remove elder age exited containers, then cause inspectcontainer report "no such container" error

**Comment by 249043822** [comment]
If we could catch the error msg"No such container" in the line below, we could ignore this error to prevent stuck in pod terminal, but we should test all the runtime could throw the similar error like "No such container", since the cri-api does not define a NotFound(404) statuscode for container not found.https://github.com/kubernetes/kubernetes/blob/9d2b361ebc7ef28f7cb75596ef40b7c239732d37/pkg/kubelet/kuberuntime/kuberuntime_container.go#L508

**Comment by rphillips** [comment]
This may be fixed by https://github.com/kubernetes/kubernetes/pull/107695

**Comment by liushi001** [comment]
> This may be fixed by #107695

I checked it, but I found our version has not include its prerequisites #104743.

**Comment by tossmilestone** [comment]
We have found the issue is still existing at 1.22.8， the #108189 doesn't fix it completely.
